### PR TITLE
Streamline Optional Dependencies

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did
- Added `dev` and `docs` optionals and organized them better
- cleaned up makefile and gitignore a bit

### Cute Animal Picture
grooming

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -7,31 +7,43 @@
...
-
-
-extras = {
-    'test': test_deps,
-    'lint': lint_deps,
+extras_require = {
+    'test': [
+        'pytest>=5.2.0,<6',
+        'pytest-cov>=2.8.1,<3',
+        'coveralls[yaml]>=1.8.2,<2',
```
Comment (iamdefinitelyahuman) [inline setup.py:14]: probably don't need this anymore since we use codecov

File: setup.py [hunk 2]
```diff
@@ -7,31 +7,43 @@
...
+        'isort>=4.2.15,<5',
+        'mypy>=0.740,<1',
+    ],
+    'docs': [
+        'recommonmark',
+        'sphinx',
+        'sphinx_rtd_theme',
+    ],
+    'dev': [
+        'ipython',
```
Comment (iamdefinitelyahuman) [inline setup.py:35]: it's a useful tool, but `ipython` as a dependency seems weird

File: setup.py [hunk 3]
```diff
@@ -7,31 +7,43 @@
...
-    'test': test_deps,
-    'lint': lint_deps,
+extras_require = {
+    'test': [
+        'pytest>=5.2.0,<6',
+        'pytest-cov>=2.8.1,<3',
+        'coveralls[yaml]>=1.8.2,<2',
+        'pytest-xdist>=1.30.0,<2',
+        'eth-tester[py-evm]>=0.3.0b1,<0.4',
+        'web3>=5.2.0,<5.3.0',
```
Comment (iamdefinitelyahuman) [inline setup.py:17]: web3 is up to `5.11.1` and we get a warning from something in it..  maybe time to bump this one as well

File: Makefile [hunk 4]
```diff
@@ -11,22 +11,46 @@ endif
...
 	@rm -fr dist/
 	@rm -fr *.egg-info
+	@rm -f vyper/vyper_git_version.txt
+
+clean-docs:
+	@echo Cleaning doc build files...
+	@rm -rf docs/_build/
+	@rm -f docs/modules.rst
+	@rm -f docs/vyper.rst
+	@rm -f docs/vyper.*.rst
```
Comment (iamdefinitelyahuman) [inline Makefile:53]: not sure these lines do anything

File: setup.py (hunk 2 above)
Comment (fubuloubu) [reply setup.py:35]: nothing under `dev` is a required dependency, just helpful

File: Makefile (hunk 4 above)
Comment (fubuloubu) [reply Makefile:53]: they do, these are intermediate build files for sphinx

**Comment by fubuloubu** [comment]
@iamdefinitelyahuman I had to make a bunch of changes for updated linter rules in d2b1461 

You might have to rebase off of it

**Comment by fubuloubu** [comment]
Hypothesis [`v5.8.6`](https://hypothesis.readthedocs.io/en/latest/changes.html#v5-8-6) changed a warning for function scoped pytest fixtures, however the warning message gives an unpredictable exception due to how it is being composed. This affects parameterized test cases only.

Slated to be fixed: https://github.com/HypothesisWorks/hypothesis/issues/2469

**Comment by fubuloubu** [comment]
Hopefully HypothesisWorks/hypothesis#2469 gets released soon, but this PR has a fixed pin that solves the problem for now.

**Pull request merged**